src/borrow_tracker/tree_borrows/tree.rs

```diff
@@ -725,27 +736,480 @@ mod commutation_tests {
+    // to two tags that are mutually foreign.
+    // The two tags are by convention named `x` and `y` throughout these tests.
+    // This includes everything except `(StrictChildAccess, StrictChildAccess)`
+    // because no tag can be a child of both.
+    #[derive(Debug, Clone, Copy, PartialEq, Eq)]
+    enum AccessRel2 {
+        ChildX,
+        ChildY,
+        Other,
+    }
+
+    impl Exhaustive for AccessRel2 {
+        fn exhaustive() -> Box<dyn Iterator<Item = Self>> {
+            use AccessRel2 as Rel;
+            Box::new(vec![Rel::ChildX, Rel::ChildY, Rel::Other].into_iter())
+        }
+    }
+
+    #[derive(Debug, Clone, Copy)]
+    struct Access2 {
+        rel2: AccessRel2,
+        kind: AccessKind,
+    }
+
+    impl Exhaustive for Access2 {
+        fn exhaustive() -> Box<dyn Iterator<Item = Self>> {
+            Box::new(
+                <(AccessRel2, AccessKind)>::exhaustive().map(|(rel2, kind)| Access2 { rel2, kind }),
+            )
+        }
+    }
+
+    #[derive(Clone, PartialEq, Eq, Hash)]
+    /// Our testing utilties include the protector bound to the state
+    /// rather than the tag. There is little room for errors since we only
+    /// do examples on single-byte locations anyway.
+    struct LocStateProt {
+        state: LocationState,
+        prot: bool,
+    }
+
+    impl LocStateProt {
+        fn perform_access(&self, kind: AccessKind, rel: AccessRelatedness) -> Result<Self, ()> {
+            let mut state = self.state;
+            let _ = state.perform_access(kind, rel, self.prot).map_err(|_| ())?;
+            Ok(Self { state, prot: self.prot })
+        }
+
+        /// End of the protector.
+        fn release(&self) -> Self {
+            let state = self.state;
+            let prot = false;
+            Self { state, prot }
+        }
+    }
+
+    impl Exhaustive for LocStateProt {
+        fn exhaustive() -> Box<dyn Iterator<Item = Self>> {
+            Box::new(
+                <(LocationState, bool)>::exhaustive().map(|(state, prot)| Self { state, prot }),
+            )
+        }
+    }
+
+    impl fmt::Display for LocStateProt {
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            write!(f, "{}", self.state)?;
+            if self.prot {
+                write!(f, " [P]")?;
+            }
+            Ok(())
+        }
+    }
+
+    #[derive(Clone, PartialEq, Eq, Hash)]
+    /// A `(LocStateProt, LocStateProt)` has more degrees of freedom
+    /// than what we want to allow for two aliasing and mutually foreign
+    /// tags (e.g. `(Active, Active)` among many others is unreachable through
+    /// normal means), so a pair of tags gets its own methods.
+    /// We'll still generate these cases exhaustively, but it prevents silly mistakes.
+    /// Pointers `x` and `y` might be used in different threads, so their protectors
+    /// could be removed in an arbitrary order.
+    struct LocStateProt2 {
+        x: LocStateProt,
+        y: LocStateProt,
```

Comment:
I'm confused, why are there two locations? There should be only one state, right, for the one location that everything points to?

Reply:
One location, but the two pointers each have their own state.
`LocStateProt` is per-pointer per-byte, so we need 2 for a 1-byte allocation with two pointers.

Reply:
Oh I see, yes that makes sense.

So the first sentence of the docs should be: for a single location, this stores the state of two pointer tags.